### PR TITLE
Update Recipe Docstrings

### PR DESCRIPTION
**Pull request by kartikayk**
#### Context
- As per title

#### Changelog
- Update docstrings
- Remove unused code from full_finetune_single_device

#### Test plan
- Run recipes

### CODE REVIEW

**Review by RdoubleA — APPROVED** [review]
Like this approach of highlighting recipe features more

File: recipes/eleuther_eval.py [hunk 1]
```diff
@@ -102,20 +102,20 @@ def _model_generate(self, *args, **kwargs):
...
 
-    Assumptions:
-        - Evaluation is launched with the Tune CLI (recommended)
-        - User has the EleutherAI eval harness installed, see https://github.com/EleutherAI/lm-evaluation-harness
+    We recommend launching evaluation using the tune CLI:
 
-    The following configs can be used to run this recipe:
-        - eleuther_eval.yaml
+        tune run eleuther_eval --config llama2_eleuther_eval \
+        tasks=["truthfulqa_mc2", "hellaswag"]
```
[the remark this answers is not in the document]
Comment (kartikayk) [reply recipes/eleuther_eval.py:118]: I think the space is helpful visually?

File: recipes/full_finetune_distributed.py [hunk 2]
```diff
@@ -38,29 +38,54 @@
...
+            activations in memory and recompute them during the backward pass. This is especially helpful
+            for larger batch sizes when you're memory constrained. But these savings in memory come at the
+            cost of training performance. In most cases training can slow-down quite a bit as a result of
+            this activation recomputation.
+
+        - Precision. Full fp32 and bf16 training are supported. Precision is controlled using the ``dtype``
+            flag. When ``dtype=bf16``, all activations, gradients and optimizer states are in bfloat16. In
+            most cases this should halve the memory footprint of full precision (fp32) training, without
+            loss in model quality (will depend on the model, training data and other settings). For
+            GPUs which do not support bfloat16, we fall back to fp32. This behavior may change in the future
```
[the remark this answers is not in the document]
Comment (kartikayk) [reply recipes/full_finetune_distributed.py:59]: Good point

File: recipes/full_finetune_distributed.py [hunk 3]
```diff
@@ -38,29 +38,54 @@
...
-    The following configs can be used to run this recipe:
-        >>> tune ls
-        RECIPE                           CONFIG
-        full_finetune_distributed        llama2/7B_full
-                                         llama2/13B_full
+    Full finetuning recipe for dense transformer-based LLMs such as Llama2. This recipe supports
+    distributed training and can be run on a single node (1 to 8 GPUs).
+
+    Features:
+        - FSDP. Supported using PyTorch's FSDP APIs. DDP is currently not supported. Training on CPU
```
[the remark this answers is not in the document]
Comment (kartikayk) [reply recipes/full_finetune_distributed.py:45]: I actually mean training on CPU

File: docs/source/examples/recipe_deepdive.rst [hunk 4]
```diff
@@ -4,11 +4,11 @@
 Training Recipe Deep-Dive
 =========================
 
-This tutorial will walk you through the design of training-recipes in torchtune.
+This deep-dive will walk you through the design of training-recipes in torchtune.
 
 .. grid:: 1
 
-    .. grid-item-card:: :octicon:`mortar-board;1em;` What you will learn
```
Comment (ebsmothers) [inline docs/source/examples/recipe_deepdive.rst:11]: (Commenting here cause it's further down in the doc) L190 still needs a change from "config tutorial" -> "config deep-dive"

File: recipes/README.md [hunk 5]
```diff
@@ -27,9 +27,9 @@ Recipes in torchtune are:
 
 &nbsp;
 
-### Architecture Optimization
+### Quantization and Sparsity
 
-torchtune integrates with `torchao`(https://github.com/pytorch-labs/ao/) for architecture optimization techniques including quantization and sparsity. Currently only some quantization techniques are integrated, see `recipes/configs/quantization.yaml` for more details.
+torchtune integrates with `torchao`(https://github.com/pytorch-labs/ao/) for architecture optimization techniques including quantization and sparsity. Currently only some quantization techniques are integrated, see the docstrings in the [quantization recipe](quantize.py) for more details.
```
Comment (ebsmothers) [inline recipes/README.md:32]: Why keep torchao as code and not just directly hyperlink here?

File: recipes/README.md [hunk 6]
```diff
@@ -38,106 +38,51 @@ tune run quantize --config quantization
...
+To run inference using a quantized model, make the following changes to the default [generation config](configs/generation.yaml)
+
+
+```yaml
+# Currently we only support torchtune checkpoints when
+# evaluating quantized models. For more details on checkpointing see
+# https://pytorch.org/torchtune/main/examples/checkpointer.html
+# Make sure to change the default checkpointer component
 checkpointer:
   _component_: torchtune.utils.FullModeltorchtuneCheckpointer
```
Comment (ebsmothers) [inline recipes/README.md:75]: similar comment down here

File: recipes/lora_finetune_single_device.py [hunk 7]
```diff
@@ -31,29 +31,57 @@
...
+
+        - Checkpointing. Model weights are checkpointed both at the end of each epoch and at the end of
+            training. Currently we checkpoint both the adapter weights (trainable params only) and the
+            complete merged weights (adapter weights added back to the base model). For more details
+            please take a look at our LoRA tutorial
+            (https://pytorch.org/torchtune/main/examples/lora_finetune.html).
+
+            Optimizer State and recipe state (seed, total_epochs, number of epochs run etc) are
+            only saved at the end of a given epoch and used in case of resuming training. Resuming
+            training is ontrolled by the ``resume_from_checkpoint`` flag. Mid-epoch checkpointing is
```
Comment (ebsmothers) [inline recipes/lora_finetune_single_device.py:75]:
```suggestion
            training is controlled by the ``resume_from_checkpoint`` flag. Mid-epoch checkpointing is
```

File: recipes/lora_finetune_single_device.py [hunk 8]
```diff
@@ -31,29 +31,57 @@
...
+            currently not supported.
+
+            For more details on the checkpointer, please take a look at
+            our checkpointer deepdive (https://pytorch.org/torchtune/main/examples/checkpointer.html).
+
+        - Logging. Terminal, WandB and TensorBoard are all supported.
```
Comment (ebsmothers) [inline recipes/lora_finetune_single_device.py:81]: Also disk, right?

File: recipes/eleuther_eval.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (RdoubleA) [reply recipes/eleuther_eval.py:118]: Don't think it adds anything, and it might be confusing if a user copies and launches this command and it fails because of the space. But this is very minor, so your call.

A separate point - why not just make these tasks default in the config?

File: recipes/lora_finetune_distributed.py [hunk 9]
```diff
@@ -40,28 +40,57 @@
...
+            GPUs which do not support bfloat16, we fall back to fp32. Mixed precision training and fp16
+            precision are currently not supported.
+
+        - Gradient Accumulation. You can simulate larger batch sizes by accumulating gradients. This is
+            controlled using the ``gradient_accumulation_steps`` flag.
+
+                Total Batch Size = batch_size * number of GPUs * gradient accumulation steps.
+
+            For example: with batch_size=1, nproc_per_node=2 and gradient_accumulation_steps=32 we get a
+            total batch size of 64.
```
Comment (ebsmothers) [inline recipes/lora_finetune_distributed.py:70]: Idk if this is the right place to point this out or not, but there are also implications for learning rate when tuning this value

File: recipes/full_finetune_single_device.py [hunk 10]
```diff
@@ -71,29 +102,24 @@ def __init__(self, cfg: DictConfig) -> None:
                 "full fp16 training is not supported with this recipe. Please use bf16 or fp32 instead."
             )
 
-        # For CUDA devices, check if the HW supports bf16 if bf16 is specified.
```
Comment (ebsmothers) [inline recipes/full_finetune_single_device.py:74]: So now we will not fail but just run in fp32 instead?

File: recipes/eleuther_eval.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ebsmothers) [reply recipes/eleuther_eval.py:118]: Did you test that it actually runs with the space? Personally I have had issues with this in the past

[the remark this answers is not in the document]
Comment (RdoubleA) [reply recipes/eleuther_eval.py:118]: Yes I've run the command with the space and it does not work

File: recipes/README.md (hunk 5 above)
Comment (kartikayk) [reply recipes/README.md:32]: oh yeh good point

File: recipes/README.md [hunk 11]
```diff
@@ -38,106 +38,51 @@ tune run quantize --config quantization
...
 
-```
-# make sure to change the checkpointer component
+```yaml
+# Currently we only support torchtune checkpoints when
+# evaluating quantized models. For more details on checkpointing see
+# https://pytorch.org/torchtune/main/examples/checkpointer.html
+# Make sure to change the default checkpointer component
 checkpointer:
   _component_: torchtune.utils.FullModeltorchtuneCheckpointer
```
[the remark this answers is not in the document]
Comment (kartikayk) [reply recipes/README.md:50]: good catch!

File: recipes/eleuther_eval.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kartikayk) [reply recipes/eleuther_eval.py:118]: Ah ok

File: recipes/lora_finetune_distributed.py (hunk 9 above)
Comment (kartikayk) [reply recipes/lora_finetune_distributed.py:70]: Do you mean LR should be appropriately tuned? Hmm yeh its a good point, but not sure if I want to point out here. I do want to capture how to think about BS when using gradient accumulation though. let me know if you feel strongly about the LR
